### PR TITLE
Update madvise call for shared heap

### PR DESCRIPTION
**Pull request by bragaigor**
If heap is created using shared memory with mmap, we must set
advice to MADV_REMOVE, becasue pages might not be immediately
freed in a successful madvise call used with MADV_DONTNEED in
shared memory case. But using MADV_REMOVE does.

Ran failing test for 1000 iterations and all passed:
https://hyc-runtimes-jenkins.swg-devops.com/view/Test_grinder/job/Grinder/5955/

Fixes: https://github.com/eclipse/openj9/issues/8385

Signed-off-by: Igor Braga <higorb1@gmail.com>

### CODE REVIEW

**Comment by bragaigor** [comment]
@genie-omr build all

**Comment by Leonardo2718** [comment]
@genie-omr build all

File: port/linux/omrvmem.c [hunk 1]
```diff
@@ -677,7 +678,14 @@ omrvmem_decommit_memory(struct OMRPortLibrary *portLibrary, void *address, uintp
...
+					advice = MADV_DONTNEED;
+					/* If heap is created using shared memory with mmap, we must set advice to MADV_REMOVE, becasue
+					 * pages might not be immediately freed in a successful madvise used with MADV_DONTNEED in
+					 * shared memory case. But using MADV_REMOVE does. */
+					if (-1 != identifier->fd) {
+						advice = MADV_REMOVE;
+					}
+					result  = (intptr_t)madvise((void *)address, (size_t) byteAmount, advice);
 				} else {
 					/* need to determine what to use in the case of shmat/shmget, till then return success */
```
Comment (amicic) [inline port/linux/omrvmem.c]: I wonder if identifier->allocator should be set to OMRPORT_VMEM_RESERVE_USED_SHM when we create heap as shared (with file descriptor), and then handle decommit in this else branche

Comment (amicic) [reply port/linux/omrvmem.c]: Also fd will be set when we create heap with huge pages, in which cases we should not try this (or any decommit).

Comment (bragaigor) [reply port/linux/omrvmem.c]: That would also work. The only case where `OMRPORT_VMEM_RESERVE_USED_SHM` is set is in the case of huge pages where we call `shmat()` to create the heap. If we switch to use `OMRPORT_VMEM_RESERVE_USED_SHM`, we would also need to update `omrvmem_free_memory()`. In theory we would need to set both `OMRPORT_VMEM_RESERVE_USED_SHM` and `OMRPORT_VMEM_RESERVE_USED_MMAP` in case of the shared heap.

Comment (amicic) [reply port/linux/omrvmem.c]: At high level, I'm ok with this change, but @dmitripivkine knows this code better. From a discussion I had with him, my proposal is no good. 

@dmitripivkine please review.  More testing was done by @bragaigor  to prove that there are no negative effects in repetitively expand & contract heap and consequently commit & decommit (using MADV_REMOVE advice) areas of heap that have been double mapped

Comment (dmitripivkine) [reply port/linux/omrvmem.c]: We never shared mmap'ed pages before, so looks like it is necessary improvement. However I don't know enough how MADV_REMOVE works exactly. I will try to find out more about it.
@pshipton @DanHeidinga Would you please take a look as well?

Comment (amicic) [reply port/linux/omrvmem.c]: in the light of my original proposal.... should we have a hybrid allocator type _MMAP_SHM?

Comment (pshipton) [reply port/linux/omrvmem.c]: +1 to Aleks' suggestion to add another allocator type to reflect shared mmap allocations. It seems using the fd to determine if shared mapping is used isn't guaranteed. There is another possible case where a /dev/zero file is used and the fd is set.

**Comment by bragaigor** [comment]
@genie-omr build all

File: port/linux/omrvmem.c [hunk 2]
```diff
@@ -1081,6 +1087,7 @@ default_pageSize_reserve_memory(struct OMRPortLibrary *portLibrary, void *addres
 	int flags = 0;
 	void *result = NULL;
 	int protectionFlags = PROT_NONE;
+	uintptr_t allocator = 0;
```
Comment (amicic) [inline port/linux/omrvmem.c]: Minor: 'Variables should always be declared in the innermost possible scope'
https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md

**Comment by amicic** [comment]
@rwy0717 please proceed with further review, but you should wait for @pshipton 's approval as well.

File: port/linux/omrvmem.c [hunk 3]
```diff
@@ -1145,7 +1151,11 @@ default_pageSize_reserve_memory(struct OMRPortLibrary *portLibrary, void *addres
 			result = NULL;
 		} else {
 			/* Update identifier and commit memory if required, else return reserved memory */
-			update_vmemIdentifier(identifier, result, result, byteAmount, mode, pageSize, OMRPORT_VMEM_PAGE_FLAG_NOT_USED, OMRPORT_VMEM_RESERVE_USED_MMAP, category, fd);
+			uintptr_t allocator = OMRPORT_VMEM_RESERVE_USED_MMAP;
+			if (useBackingSharedFile) {
+				allocator = OMRPORT_VMEM_RESERVE_USED_MMAP_SHM;
+			}
+			update_vmemIdentifier(identifier, result, result, byteAmount, mode, pageSize, OMRPORT_VMEM_PAGE_FLAG_NOT_USED, allocator, category, fd);
```
Comment (pshipton) [inline port/linux/omrvmem.c]: What about the double mapping at https://github.com/eclipse/omr/blob/82079d3d56995cfb3b2e330f35f4b3d7faa4a03b/port/linux/omrvmem.c#L1212, should it be using OMRPORT_VMEM_RESERVE_USED_MMAP_SHM as well?

Comment (bragaigor) [reply port/linux/omrvmem.c]: The contiguous representation of the arraylet is private but the arraylets themselves (with double mapping) are shared, so your suggestion is probably the right way to do it. I'll update it

**Comment by bragaigor** [comment]
@genie-omr build all

**Comment by amicic** [comment]
Hm, not an expert, but I'm more inclined to say it's not needed.

File: port/linux/omrvmem.c (hunk 3 above)
Comment (amicic) [reply port/linux/omrvmem.c]: As I was saying in another comment, I did not think it's needed since would not normally try to decommit through this double mapped contiguous memory. 

But for some other potential users of this API that might try to decommit, I understand the concern. 

So, we can temporarily modify GC code and decommit memory (using contiguous address) just prior to unmapping it, and compare footprint (at the end of GC or end of a test) when allocator is as is, MMAP, hence decommited with DONTNEED advice vs MMAP_SHM, hence decommited through REMOVE advice.

Comment (bragaigor) [reply port/linux/omrvmem.c]: Some observations that I noticed while testing this fix. We don't actually need the commit part when double mapping an arraylet at:
https://github.com/eclipse/omr/blob/82079d3d56995cfb3b2e330f35f4b3d7faa4a03b/port/linux/omrvmem.c#L1214 
through line 1223 because that's already taken care by `get_protectionBits(mode)` at:
https://github.com/eclipse/omr/blob/82079d3d56995cfb3b2e330f35f4b3d7faa4a03b/port/linux/omrvmem.c#L1193
I tested and behaviour stayed the same if we remove the commit part.
And in regards to double mapped arraylets we do not decommit them anywhere, all we do is call the `vmem_free_memory()` API whenever they become garbage to free them, which doesn't decommit the memory. But the question still remains whether we should use `OMRPORT_VMEM_RESERVE_USED_MMAP_SHM` or `OMRPORT_VMEM_RESERVE_USED_MMAP`. The code as it stands wouldn't change in behaviour if either is used. Probably the best approach would be removing the commit from double mapping and update to `_MMAP_SHM` since the heap is using that same flag and double mapped arrays are shared through mmap?

Comment (pshipton) [reply port/linux/omrvmem.c]: Yes, it seem more correct to use `_MMAP_SHM` even if the GC doesn't decommit that way. Something else could use the API and call decommit.

Comment (pshipton) [reply port/linux/omrvmem.c]: About the commit, same reasoning. It's not only about what the GC needs. The APIs should operate correctly even if used for a different purpose. What if a user wanted to double map executable memory, then the commit is needed.

Comment (bragaigor) [reply port/linux/omrvmem.c]: On top of that I would probably add a call to `vmem_decommit_memory()` just before `vmem_free_memory()` whenever we are freeing the contiguous representation of a double mapped arraylet. In this case I would still choose the flag `OMRPORT_VMEM_RESERVE_USED_MMAP_SHM` because that forces `madvise()` to be called with `MADV_REMOVE` which tells the OS that we won't need that memory again (releases memory back to OS immediately). That's safe because the contiguous representation of the arraylet is created as an anonymous private memory through mmap (the OS chooses where to put) set by the flags here:
https://github.com/eclipse/omr/blob/82079d3d56995cfb3b2e330f35f4b3d7faa4a03b/port/linux/omrvmem.c#L1186
Hence, we do not know where the OS is going to put the contiguous representation of the next arraylet, so it's better to give the memory back to the OS so that the OS has that memory available for future usage (e.g. allocating more contiguous representation of arraylets)

Comment (bragaigor) [reply port/linux/omrvmem.c]: Correct, but looking at the API it seems the commit call here is redundant because we use `mmap` to create the contiguous representation of the arraylets with the same flags and all the commit call is doing is calling `mprotect()` to change the protection for the calling process's memory page, and the protection is already set by the `protectionFlags` here when we `mmap` it:
https://github.com/eclipse/omr/blob/82079d3d56995cfb3b2e330f35f4b3d7faa4a03b/port/linux/omrvmem.c#L1194
I'm probably missing something?

Comment (pshipton) [reply port/linux/omrvmem.c]: ok, makes sense.

Comment (bragaigor) [reply port/linux/omrvmem.c]: I did ran some tests locally just to make sure the above is correct. I ran a few tests from my local C program where I simulate a heap with a few arraylet leaves in it. It looks like the madvise()/decommit call with `MADV_DONTNEED` is not sufficient to free the contiguous representation of the arraylet and `MADV_REMOVE` does free that memory:

Case 1:

1. Create arraylet leaves
2. Call double mapping (create contiguous representation of arraylet)
3. Call `madvise` with `MADV_DONTNEED` on the contiguous address
4. Check free physical memory <—— Memory NOT freed
5. Call `munmap` on the contiguous address
6. Check free physical memory <—— Memory NOT freed
Here physical memory is not freed with neither `munmap` or `madvise` call

Case 2:

1. Create arraylet leaves
2. Call double mapping (create contiguous representation of arraylet)
3. Call `madvise` with `MADV_REMOVE` on the contiguous address
4. Check free physical memory <—— Most memory freed
5. Call `munmap` on the contiguous address
6. Check free physical memory <—— All Memory freed

Most memory is freed in the `madvise` call but some remnants are freed after `munmap` call. This shows that `OMRPORT_VMEM_RESERVE_USED_MMAP_SHM` is the correct flag to use when dealing with double mapped arraylet lets.

Comment (amicic) [reply port/linux/omrvmem.c]: Thanks, Igor, I confirm I'm ok too with extra addition that Peter suggested.
@rwy0717 please proceed with further steps

**Comment by mstoodle** [comment]
@rwy0717 I think you're more knowledgeable in this area than other committers (unless @charliegracie gets interested :) ) Would you please be committer here? It seems like everybody else has kindly already marked their approvals for you to consider.
